packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/RealmAny.kt

```diff
@@ -0,0 +1,329 @@
+package io.realm.kotlin.types
+
+import io.realm.kotlin.dynamic.DynamicMutableRealmObject
+import io.realm.kotlin.dynamic.DynamicRealmObject
+import io.realm.kotlin.internal.RealmAnyImpl
+import io.realm.kotlin.query.RealmQuery
+import org.mongodb.kbson.BsonObjectId
+import org.mongodb.kbson.Decimal128
+import kotlin.reflect.KClass
+
+/**
+ * `RealmAny` is used to represent a polymorphic Realm value.
+ *
+ * At any particular moment an instance of this class stores a definite value of a definite type.
+ * If, for instance, that is a `Double` value, you may call [asDouble] to extract that value. You
+ * may call [type] to discover what type of value is currently stored. Calling `asDouble` on an
+ * instance that does not store a `Double` value would raise an [IllegalStateException].
+ *
+ * `RealmAny` behaves like a value type on all the supported types except on Realm objects. It means
+ * that Realm will not persist any change to the actual `RealmAny` value. If a `RealmAny` holds a
+ * [RealmObject], it just holds the reference to it, not a copy of the object. Because `RealmAny`
+ * instances are immutable, a new instance is needed to update a `RealmAny` attribute.
+ * ```
+ *      anObject.realmAnyField = RealmAny.create(42.0)
+ *      anObject.realmAnyField = RealmAny.create("Hello")
+ *      anObject.realmAnyField = RealmAny.create(MyRealmObject())
```

Comment:
I think that we should add an explicit `NOTE` that we do not support embedded objects in `RealmAny`. I know that it is given by the type system, but would rule out any confusion.

Reply:
I wrote a comment about it here https://github.com/realm/realm-kotlin/blob/27ac20abe66715808b7628d7f138aa6fc8f396d6/packages/library-base/src/commonMain/kotlin/io/realm/kotlin/types/RealmAny.kt#L67 but perhaps it's not too visible.